staging/src/k8s.io/apiserver/pkg/util/configmetrics/info_collector.go

```diff
@@ -0,0 +1,125 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configmetrics
+
+import (
+	"sync/atomic"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/utils/ptr"
+)
+
+// HashProvider is an interface for getting the current config hash values
+type HashProvider interface {
+	GetCurrentHashes() map[string]string
+}
+
+// AtomicHashProvider implements HashProvider using atomic operations for thread-safe access
+type AtomicHashProvider struct {
+	hashes map[string]*atomic.Pointer[string]
+}
+
+// NewAtomicHashProvider creates a new atomic hash provider with the specified label names
+func NewAtomicHashProvider(labelNames ...string) *AtomicHashProvider {
+	p := &AtomicHashProvider{
+		hashes: make(map[string]*atomic.Pointer[string]),
+	}
+	// Initialize with empty strings for each label
+	for _, name := range labelNames {
+		atomicPtr := &atomic.Pointer[string]{}
+		atomicPtr.Store(ptr.To(""))
+		p.hashes[name] = atomicPtr
+	}
+	return p
+}
+
+func (h *AtomicHashProvider) GetCurrentHashes() map[string]string {
+	result := make(map[string]string)
+	for name, ptr := range h.hashes {
+		if valuePtr := ptr.Load(); valuePtr != nil {
+			result[name] = *valuePtr
+		} else {
+			result[name] = ""
+		}
+	}
+	return result
+}
+
+func (h *AtomicHashProvider) SetHash(name, value string) {
+	if atomicPtr, exists := h.hashes[name]; exists {
+		atomicPtr.Store(ptr.To(value))
+	}
+}
+
+func (h *AtomicHashProvider) SetHashes(values map[string]string) {
+	for name, value := range values {
+		h.SetHash(name, value)
+	}
+}
+
+func (h *AtomicHashProvider) Reset() {
+	for _, atomicPtr := range h.hashes {
+		atomicPtr.Store(ptr.To(""))
+	}
+}
+
+// NewConfigInfoCustomCollector creates a custom collector for config hash info metrics.
+// This eliminates the need for state management and locks by collecting metrics on demand.
+func NewConfigInfoCustomCollector(desc *metrics.Desc, hashProvider HashProvider, labelNames []string) metrics.StableCollector {
+	return &configInfoCustomCollector{
+		desc:         desc,
+		hashProvider: hashProvider,
+		labelNames:   labelNames,
+	}
+}
+
+type configInfoCustomCollector struct {
+	metrics.BaseStableCollector
+	desc         *metrics.Desc
+	hashProvider HashProvider
+	labelNames   []string
+}
+
+var _ metrics.StableCollector = &configInfoCustomCollector{}
+
+func (c *configInfoCustomCollector) DescribeWithStability(ch chan<- *metrics.Desc) {
+	ch <- c.desc
+}
+
+func (c *configInfoCustomCollector) CollectWithStability(ch chan<- metrics.Metric) {
+	hashes := c.hashProvider.GetCurrentHashes()
+
+	// Only emit metric if all hash values are non-empty
+	allNonEmpty := true
+	for _, value := range hashes {
+		if value == "" {
+			allNonEmpty = false
+			break
+		}
+	}
+
+	if allNonEmpty && len(hashes) > 0 {
```

Comment:
I think this `if` can be dropped once you have the above change because `len(hashes) > 0` should always be `> 0`.